File: rust/drift-rs/crates/src/math/tiers.rs
use crate::types::{AssetTier, ContractTier};

/// Extension trait: numeric ordering for drift's `ContractTier` (orphan rule).
pub trait ContractTierExt {
    fn to_number(&self) -> u8;
}
impl ContractTierExt for ContractTier {
    fn to_number(&self) -> u8 {
        match self {
            ContractTier::A => 0,
            ContractTier::B => 1,
            ContractTier::C => 2,
            ContractTier::Speculative => 3,
            ContractTier::HighlySpeculative => 4,
            ContractTier::Isolated => 5,
        }
    }
}

/// Extension trait: numeric ordering for drift's `AssetTier` (orphan rule).
pub trait AssetTierExt {
    fn to_number(&self) -> u8;
}
impl AssetTierExt for AssetTier {
    fn to_number(&self) -> u8 {
        match self {
            AssetTier::Collateral => 0,
            AssetTier::Protected => 1,
            AssetTier::Cross => 2,
            AssetTier::Isolated => 3,
            AssetTier::Unlisted => 4,
        }
    }
}

pub fn perp_tier_is_as_safe_as(perp_tier: u8, other_perp_tier: u8, other_spot_tier: u8) -> bool {
    let as_safe_as_perp = perp_tier <= other_perp_tier;
    let as_safe_as_spot = other_spot_tier == 4 || (other_spot_tier >= 2 && perp_tier <= 2);
    as_safe_as_spot && as_safe_as_perp
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_tier_safety_comparison() {
        let tier_a = ContractTier::A;
        let tier_b = ContractTier::B;
        let spot_cross = AssetTier::Cross;

        assert!(perp_tier_is_as_safe_as(
            tier_a.to_number(),
            tier_b.to_number(),
            spot_cross.to_number()
        ));

        assert!(!perp_tier_is_as_safe_as(
            tier_b.to_number(),
            tier_a.to_number(),
            spot_cross.to_number()
        ));
    }
}

File: packages/sdk/src/math/tiers.ts
import { isVariant, PerpMarketAccount, SpotMarketAccount } from '../types';

/**
 * Maps a perp market's `contractTier` to an ordinal safety rank, lower is safer. Matches the
 * declaration order of the Rust `ContractTier` enum (which derives `Ord` from declaration order,
 * used by `ContractTier::is_as_safe_as_contract`'s `self <= other`).
 *
 * @param {PerpMarketAccount} perpMarket - The perp market account
 * @return {number} `0` (A, safest) through `5` (Isolated, riskiest); `4` = HighlySpeculative
 */
export function getPerpMarketTierNumber(perpMarket: PerpMarketAccount): number {
	if (isVariant(perpMarket.contractTier, 'a')) {
		return 0;
	} else if (isVariant(perpMarket.contractTier, 'b')) {
		return 1;
	} else if (isVariant(perpMarket.contractTier, 'c')) {
		return 2;
	} else if (isVariant(perpMarket.contractTier, 'speculative')) {
		return 3;
	} else if (isVariant(perpMarket.contractTier, 'highlySpeculative')) {
		return 4;
	} else {
		return 5;
	}
}

/**
 * Maps a spot market's `assetTier` to an ordinal safety rank, lower is safer. Matches the
 * declaration order of the Rust `AssetTier` enum, used by `ContractTier::is_as_safe_as_asset`.
 *
 * @param {SpotMarketAccount} spotMarket - The spot market account
 * @return {number} `0` (Collateral, safest) through `4` (Unlisted, riskiest); `5` is unreachable
 *   (falls through only if `assetTier` matches none of the known variants)
 */
export function getSpotMarketTierNumber(spotMarket: SpotMarketAccount): number {
	if (isVariant(spotMarket.assetTier, 'collateral')) {
		return 0;
	} else if (isVariant(spotMarket.assetTier, 'protected')) {
		return 1;
	} else if (isVariant(spotMarket.assetTier, 'cross')) {
		return 2;
	} else if (isVariant(spotMarket.assetTier, 'isolated')) {
		return 3;
	} else if (isVariant(spotMarket.assetTier, 'unlisted')) {
		return 4;
	} else {
		return 5;
	}
}

/**
 * True if a perp market's tier is at least as safe as both a reference perp tier and a reference
 * spot tier, mirroring `ContractTier::is_as_safe_as`. Used to gate cross-margining: a position in
 * a market riskier than the account's other collateral/positions can force isolated margin.
 * A perp tier is "as safe as" a spot tier if the spot tier is Unlisted (anything beats Unlisted);
 * otherwise, if the spot tier is Cross or Isolated, the perp tier must be C-or-safer (tiers 0-2).
 *
 * @param {number} perpTier - This market's tier number, from `getPerpMarketTierNumber`
 * @param {number} otherPerpTier - The reference perp tier number to compare against
 * @param {number} otherSpotTier - The reference spot tier number to compare against, from
 *   `getSpotMarketTierNumber`
 * @return {boolean} Whether `perpTier` is as safe as both references
 */
export function perpTierIsAsSafeAs(
	perpTier: number,
	otherPerpTier: number,
	otherSpotTier: number
): boolean {
	const asSafeAsPerp = perpTier <= otherPerpTier;
	const asSafeAsSpot =
		otherSpotTier === 4 || (otherSpotTier >= 2 && perpTier <= 2);
	return asSafeAsSpot && asSafeAsPerp;
}
